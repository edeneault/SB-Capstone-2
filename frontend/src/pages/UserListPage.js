import React, { useEffect } from "react";
import { LinkContainer } from "react-router-bootstrap";
import { Container, Table, Button } from "react-bootstrap";
import { useDispatch, useSelector } from "react-redux";
import Loader from "../components/Loader";
import { usersList, deleteUser } from "../actions/userActions";

import { toast } from "react-toastify";
import "react-toastify/dist/ReactToastify.css";

toast.configure();

const UserListScreen = ({ history }) => {
  const dispatch = useDispatch();

  const listUsers = useSelector((state) => state.usersList);
  const { loading, error, users } = listUsers;

  const userLogin = useSelector((state) => state.userLogin);
  const { userInfo } = userLogin;

  const userDelete = useSelector((state) => state.userDelete);
  const { success: successDelete } = userDelete;

  useEffect(() => {
    if (userInfo && userInfo.isAdmin) {
      dispatch(usersList());
    } else {
      history.push("/login");
    }
  }, [dispatch, history, userInfo, successDelete]);

  const deleteHandler = (id) => {
    if (window.confirm("Are you sure?")) {
      dispatch(deleteUser(id));
    }
  };
  return (
    <>
      <Container fluid className='px-5'>
        <h1 className='page-title slide-in'>My Users</h1>
        {loading ? (
          <Loader />
        ) : error ? (
          toast(`${error}`, { type: "error" })
        ) : (
          <Table bordered hover responsive className='table-sm'>
            <thead>
              <tr className='bg-black text-light'>
                <th>ID</th>
                <th>NAME</th>
                <th>EMAIL</th>
                <th>ADMIN</th>
                <th>ACTIONS</th>
              </tr>
            </thead>
            <tbody>
              {users.map((user) => (
                <tr key={user._id} className='align-middle'>
                  <td>{user._id}</td>
                  <td>{user.name}</td>
                  <td>
                    <a href={`mailto:${user.email}`}>{user.email}</a>
                  </td>
                  <td className='text-center'>
                    {user.isAdmin ? (
                      <i
                        className='fas fa-check'
                        style={{ color: "green" }}
                      ></i>
                    ) : (
                      <i className='fas fa-times' style={{ color: "red" }}></i>
                    )}
                  </td>
                  <td>
                    <LinkContainer to={`/admin/user/${user._id}/edit`}>
                      <Button variant='light' className='btn-sm edit m-1'>
                        <i className='fas fa-edit'></i>
                      </Button>
                    </LinkContainer>
                    <Button
                      className='btn-sm trash m-1'
                      onClick={() => deleteHandler(user._id)}
                    >
                      <i className='fas fa-trash'></i>
                    </Button>
                  </td>
                </tr>
              ))}
            </tbody>
          </Table>
        )}
      </Container>
    </>
  );
};

export default UserListScreen;
