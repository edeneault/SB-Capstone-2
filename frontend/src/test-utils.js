import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";

// create a customRender that wraps the UI in a memory Router

const MemoryRouterWithInitialRoutes = ({ children, initialRoutes }) => {
  return <MemoryRouter initialEntries={initialRoutes}>{children}</MemoryRouter>;
};

const customRender = (ui, options) => {
  const initialRoutes =
    options && options.initialRoutes ? options.initialRoutes : ["/"];

  return render(ui, {
    wrapper: (args) =>
      MemoryRouterWithInitialRoutes({
        ...args,
        initialRoutes,
      }),
    ...options,
  });
};

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
