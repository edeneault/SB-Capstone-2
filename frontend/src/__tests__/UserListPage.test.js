import React from "react";
import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";
import thunk from "redux-thunk";
import { initialState, render } from "../test-utils";
import UserListPage from "../pages/UserListPage.js";
import { createMemoryHistory, createLocation } from "history";

const mockStore = configureMockStore([thunk]);

describe("test", () => {
  const history = createMemoryHistory();
  const path = `/admin/usersList`;

  const match = {
    params: {
      pageNumber: 1,
    },
    url: path,
  };
  const location = createLocation(match.url);

  test("renders UserListPage without crashing", () => {
    const store = mockStore(initialState);

    const { getByText } = render(
      <Provider store={store}>
        <UserListPage history={history} match={match} />
      </Provider>,
      { initialRoutes: [`/admin/usersList`] },
    );
    // screen.debug();
    expect(getByText(/My Users/i)).toBeInTheDocument();
  });

  test("matches snapshot", function () {
    const store = mockStore(initialState);

    const { asFragment } = render(
      <Provider store={store}>
        <UserListPage history={history} match={match} />
      </Provider>,
      { initialRoutes: [`/admin/usersList`] },
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
