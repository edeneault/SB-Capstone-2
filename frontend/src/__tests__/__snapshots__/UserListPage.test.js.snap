// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="px-5 pt-3 container-fluid"
  >
    <h1
      class="page-title slide-in"
    >
      My Users
    </h1>
    <div
      class="table-responsive"
    >
      <table
        class="table-sm table table-bordered table-hover"
      >
        <thead>
          <tr
            class="bg-black text-light"
          >
            <th>
              ID
            </th>
            <th>
              NAME
            </th>
            <th>
              EMAIL
            </th>
            <th>
              ADMIN
            </th>
            <th>
              ACTIONS
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            class="align-middle"
          >
            <td>
              6125d04db949259bdec4de34
            </td>
            <td>
              AdminName
            </td>
            <td>
              <a
                href="mailto:admin@example.com"
              >
                admin@example.com
              </a>
            </td>
            <td
              class="text-center"
            >
              <i
                class="fas fa-check"
                style="color: green;"
              />
            </td>
            <td>
              <a
                class="btn-sm edit m-1 btn btn-light"
                href="/admin/user/6125d04db949259bdec4de34/edit"
              >
                <i
                  class="fas fa-edit"
                />
              </a>
              <button
                class="btn-sm trash m-1 btn btn-primary"
                type="button"
              >
                <i
                  class="fas fa-trash"
                />
              </button>
            </td>
          </tr>
          <tr
            class="align-middle"
          >
            <td>
              6125d04db949259bdec4de36
            </td>
            <td>
              Jane Doe
            </td>
            <td>
              <a
                href="mailto:jane@example.com"
              >
                jane@example.com
              </a>
            </td>
            <td
              class="text-center"
            >
              <i
                class="fas fa-times"
                style="color: red;"
              />
            </td>
            <td>
              <a
                class="btn-sm edit m-1 btn btn-light"
                href="/admin/user/6125d04db949259bdec4de36/edit"
              >
                <i
                  class="fas fa-edit"
                />
              </a>
              <button
                class="btn-sm trash m-1 btn btn-primary"
                type="button"
              >
                <i
                  class="fas fa-trash"
                />
              </button>
            </td>
          </tr>
          <tr
            class="align-middle"
          >
            <td>
              61328224147ba5000402c30a
            </td>
            <td>
              Ashley Dragan
            </td>
            <td>
              <a
                href="mailto:ashleyndragan@gmail.com"
              >
                ashleyndragan@gmail.com
              </a>
            </td>
            <td
              class="text-center"
            >
              <i
                class="fas fa-times"
                style="color: red;"
              />
            </td>
            <td>
              <a
                class="btn-sm edit m-1 btn btn-light"
                href="/admin/user/61328224147ba5000402c30a/edit"
              >
                <i
                  class="fas fa-edit"
                />
              </a>
              <button
                class="btn-sm trash m-1 btn btn-primary"
                type="button"
              >
                <i
                  class="fas fa-trash"
                />
              </button>
            </td>
          </tr>
          <tr
            class="align-middle"
          >
            <td>
              613d79e01b6ca0000427aa9b
            </td>
            <td>
              John Doe
            </td>
            <td>
              <a
                href="mailto:g@example.com"
              >
                g@example.com
              </a>
            </td>
            <td
              class="text-center"
            >
              <i
                class="fas fa-check"
                style="color: green;"
              />
            </td>
            <td>
              <a
                class="btn-sm edit m-1 btn btn-light"
                href="/admin/user/613d79e01b6ca0000427aa9b/edit"
              >
                <i
                  class="fas fa-edit"
                />
              </a>
              <button
                class="btn-sm trash m-1 btn btn-primary"
                type="button"
              >
                <i
                  class="fas fa-trash"
                />
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;
